File: ruby/lib/verse.rb
# frozen_string_literal: true

class Verse
  RESET_TEXT      = "99 bottles"
  ONE_BOTTLE      = "1 bottle"
  NO_MORE_BOTTLES = "No more bottles"

  def initialize(number)
    @number = number
  end

  def first_sentence
    case number
    when 0 then NO_MORE_BOTTLES
    when 1 then ONE_BOTTLE
    else
      number_with_bottles do |bottles|
        "#{number} #{bottles}"
      end
    end
  end

  def second_sentence
    case number
    when 0 then NO_MORE_BOTTLES.downcase
    when 1 then ONE_BOTTLE
    else
      number_with_bottles do |bottles|
        "#{number} #{bottles}"
      end
    end
  end

  def third_sentence
    case number
    when 0 then "Go to the store and buy some more"
    when 1 then "Take it down and pass it around"
    else "Take one down and pass it around"
    end
  end

  def last_sentence
    minus = number - 1
    case minus
    when -1 then RESET_TEXT
    when 0 then NO_MORE_BOTTLES.downcase
    when 1 then ONE_BOTTLE
    else
      number_with_bottles do |bottles|
        "#{minus} #{bottles}"
      end
    end
  end

  private
  attr_reader :number

  def number_with_bottles
    yield "bottles"
  end
end

File: python/lib/verse.py
class Verse:
    def first_sentence(self, number):
        if number == 0:
            return self.no_more_bottles(capitalize=True)
        elif number == 1:
            return self.one_bottle()
        else:
            return f"{number} bottles"

    def second_sentence(self, number):
        if number == 0:
            return self.no_more_bottles()
        elif number == 1:
            return self.one_bottle()
        else:
            return f"{number} bottles"

    def third_sentence(self, number):
        if number == 0:
            return "Go to the store and buy some more"
        elif number == 1:
            return "Take it down and pass it around"
        else:
            return "Take one down and pass it around"

    def last_sentence(self, number):
        minus = number - 1
        if minus < 0:
            return "99 bottles"
        elif minus == 0:
            return self.no_more_bottles()
        elif minus == 1:
            return self.one_bottle()
        else:
            return f"{minus} bottles"

    def no_more_bottles(self, capitalize=False):
        text= "no more bottles"

        if capitalize:
            return text.capitalize()
        else:
            return text

    def one_bottle(self):
        return "1 bottle"
